service/history/archivalQueueFactory.go

```diff
@@ -0,0 +1,140 @@
+		queues.NamespacePrioritySchedulerOptions{
+			WorkerCount:            params.Config.ArchivalProcessorSchedulerWorkerCount,
+			ActiveNamespaceWeights: params.Config.ArchivalProcessorSchedulerRoundRobinWeights,
+			// we don't need standby weights because we only run in the active cluster
+			EnableRateLimiter:           params.Config.TaskSchedulerEnableRateLimiter,
+			MaxDispatchThrottleDuration: HostSchedulerMaxDispatchThrottleDuration,
+		},
+		params.NamespaceRegistry,
+		params.SchedulerRateLimiter,
+		params.TimeSource,
+		params.MetricsHandler.WithTags(metrics.OperationTag(metrics.OperationArchivalQueueProcessorScope)),
+		params.Logger,
+	)
+	return &archivalQueueFactory{
+		archivalQueueFactoryParams: params,
+		QueueFactoryBase: QueueFactoryBase{
+			HostScheduler:        hostScheduler,
+			HostPriorityAssigner: queues.NewPriorityAssigner(),
+			HostRateLimiter: NewQueueHostRateLimiter(
+				params.Config.ArchivalProcessorMaxPollHostRPS,
+				params.Config.PersistenceMaxQPS,
+				archivalQueuePersistenceMaxRPSRatio,
+			),
+			HostReaderRateLimiter: queues.NewReaderPriorityRateLimiter(
+				NewHostRateLimiterRateFn(
+					params.Config.ArchivalProcessorMaxPollHostRPS,
+					params.Config.PersistenceMaxQPS,
+					archivalQueuePersistenceMaxRPSRatio,
+				),
+				params.Config.QueueMaxReaderCount(),
+			),
+		},
+	}
+}
+
+func (f *archivalQueueFactory) CreateQueue(
+	shard shard.Context,
+	workflowCache workflow.Cache,
+) queues.Queue {
+	logger := log.With(shard.GetLogger(), tag.ComponentArchivalQueue)
+
+	executor := newArchivalQueueTaskExecutor(f.archiver, shard, workflowCache, f.MetricsHandler, f.Logger)
+
+	return queues.NewImmediateQueue(
+		shard,
+		tasks.CategoryArchival,
+		f.HostScheduler,
+		f.HostPriorityAssigner,
+		executor,
+		&queues.Options{
+			ReaderOptions: queues.ReaderOptions{
+				BatchSize:            f.Config.ArchivalTaskBatchSize,
+				MaxPendingTasksCount: f.Config.QueuePendingTaskMaxCount,
+				PollBackoffInterval:  f.Config.ArchivalProcessorPollBackoffInterval,
+			},
+			MonitorOptions: queues.MonitorOptions{
+				PendingTasksCriticalCount:   f.Config.QueuePendingTaskCriticalCount,
+				ReaderStuckCriticalAttempts: f.Config.QueueReaderStuckCriticalAttempts,
+				SliceCountCriticalThreshold: f.Config.QueueCriticalSlicesCount,
+			},
+			MaxPollRPS:                          f.Config.ArchivalProcessorMaxPollRPS,
+			MaxPollInterval:                     f.Config.ArchivalProcessorMaxPollInterval,
+			MaxPollIntervalJitterCoefficient:    f.Config.ArchivalProcessorMaxPollIntervalJitterCoefficient,
+			CheckpointInterval:                  f.Config.ArchivalProcessorUpdateAckInterval,
+			CheckpointIntervalJitterCoefficient: f.Config.ArchivalProcessorUpdateAckIntervalJitterCoefficient,
+			MaxReaderCount:                      f.Config.QueueMaxReaderCount,
+			TaskMaxRetryCount: func() int {
+				// archival queue task will be retried forever
+				return math.MaxInt32
+			},
```

Comment:
This is actually critical count for logging/metrics, plz don't use max int32.

Reply:
Fixed